provisioning/device/src/Transports/Mqtt/ProvisioningTransportHandlerMqtt.cs

```diff
@@ -61,279 +64,315 @@ public class ProvisioningTransportHandlerMqtt : ProvisioningTransportHandler
-                        operation = await ProvisionOverWssUsingSymmetricKeyAsync(message, cancellationToken).ConfigureAwait(false);
-                    }
-                    else
-                    {
-                        throw new NotSupportedException($"Not supported {nameof(TransportProtocol)} value: {TransportProtocol}");
-                    }
-                }
-                else
-                {
-                    if (Logging.IsEnabled)
-                        Logging.Error(this, $"Invalid {nameof(AuthenticationProvider)} type.");
+                await mqttClient.ConnectAsync(mqttClientOptionsBuilder.Build(), cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                throw new ProvisioningTransportException("Failed to open the MQTT connection", e);
+            }
 
-                    throw new NotSupportedException(
-                        $"{nameof(message.Authentication)} must be of type {nameof(AuthenticationProviderX509)} or {nameof(AuthenticationProviderSymmetricKey)}");
-                }
+            await SubscribeToRegistrationResponseMessagesAsync(mqttClient, cancellationToken).ConfigureAwait(false);
+
+            RegistrationOperationStatus registrationStatus = await PublishRegistrationRequest(mqttClient, provisioningRequest, cancellationToken).ConfigureAwait(false);
+
+            if (Logging.IsEnabled)
+                Logging.Info(this, "Successfully sent the initial registration request, now polling until provisioning has finished.");
+
+            return await PollUntilProvisionigFinishes(mqttClient, registrationStatus.OperationId, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<RegistrationOperationStatus> PublishRegistrationRequest(IMqttClient mqttClient, ProvisioningTransportRegisterRequest provisioningRequest, CancellationToken cancellationToken)
+        {
+            DeviceRegistration registrationRequest = new DeviceRegistration()
+            {
+                Payload = new JRaw(provisioningRequest.Payload),
+            };
 
-                return operation.RegistrationState;
+            byte[] payload = new byte[0];
+
+            if (provisioningRequest.Payload != null)
+            {
+                string requestString = JsonConvert.SerializeObject(registrationRequest);
+                payload = Encoding.UTF8.GetBytes(requestString);
             }
-            catch (Exception ex) when (!(ex is ProvisioningTransportException))
+
+            int packetId = GetNextPacketId();
+            string registrationTopic = string.Format(CultureInfo.InvariantCulture, RegisterTopic, packetId);
+            var message = new MqttApplicationMessageBuilder()
+                .WithPayload(payload)
+                .WithTopic(registrationTopic)
+                .Build();
+
+            _startProvisioningRequestStatusSource = new TaskCompletionSource<RegistrationOperationStatus>();
```

Comment:
This is a class field. Do we need to worry about overwriting a previous assignment?

Reply:
I plan on making a given DPS device client's ```RegisterAsync``` call atomic. That way we can do things like this. There isn't a compelling reason why a user would want to initiate two registrations for the same device after all